planetiler-custommap/src/main/java/com/onthegomap/planetiler/custommap/ConfiguredFeature.java

```diff
@@ -0,0 +1,317 @@
+
+    BiFunction<SourceFeature, Object, Byte> attributeZoomProducer =
+      attributeZoomThreshold(minTileCoverage, attributeMinZoom, minZoomByValue);
+
+    if (attributeZoomProducer != null) {
+      return (sf, f) -> {
+        if (attributeTest.evaluate(sf)) {
+          Object value = attributeValueProducer.apply(sf);
+          f.setAttrWithMinzoom(tagKey, value, attributeZoomProducer.apply(sf, value));
+        }
+      };
+    }
+
+    return (sf, f) -> {
+      if (attributeTest.evaluate(sf)) {
+        f.setAttr(tagKey, attributeValueProducer.apply(sf));
+      }
+    };
+  }
+
+  /**
+   * Generates a factory method which creates a {@link Feature} from a {@link FeatureCollector} of the appropriate
+   * geometry type.
+   * 
+   * @param layerName - name of the layer
+   * @param type      - type of geometry
+   * @return geometry factory method
+   */
+  private Function<FeatureCollector, Feature> geometryMapFeature(String layerName, GeometryType type) {
+    return switch (type) {
+      case POLYGON -> fc -> fc.polygon(layerName);
+      case LINE -> fc -> fc.line(layerName);
+      case POINT -> fc -> fc.point(layerName);
+      default -> throw new IllegalArgumentException("Unhandled geometry type " + type);
+    };
+  }
+
+  /**
+   * Generates a test for whether a source feature is of the correct geometry to be included in the tile.
+   * 
+   * @param type type of geometry
+   * @return geometry test method
+   */
+  private static Predicate<SourceFeature> geometryTest(GeometryType type) {
+    return switch (type) {
+      case POLYGON -> SourceFeature::canBePolygon;
+      case LINE -> SourceFeature::canBeLine;
+      case POINT -> SourceFeature::isPoint;
+      default -> throw new IllegalArgumentException("Unhandled geometry type " + type);
+    };
+  }
+
+  /**
+   * Determines whether to include a source feature in the tiles.
+   * 
+   * @param sourceFeature - source feature
+   * @return true if the feature will be rendered
+   */
+  public boolean includeWhen(SourceFeature sourceFeature) {
+    //Is this from the right source?
+    if (!sources.contains(sourceFeature.getSource())) {
+      return false;
+    }
+
+    //Is this the right type of geometry?
+    if (!geometryTest.test(sourceFeature)) {
+      return false;
+    }
+
+    //Check for matching tags
+    return tagTest.evaluate(sourceFeature);
+  }
+
+  /**
+   * Generates a tile feature based on a source feature.
+   * 
+   * @param sourceFeature - input source feature
+   * @param features      - output rendered feature collector
+   */
+  public void processFeature(SourceFeature sourceFeature, FeatureCollector features) {
+
+    var minZoom = zoomOverride.getOrElse(sourceFeature, featureMinZoom);
+
+    Feature f = geometryFactory.apply(features)
+      .setMinZoom(minZoom);
+    if (featureMaxZoom != null) {
+      f.setMaxZoom(featureMaxZoom);
+    }
+    attributeProcessors.forEach(p -> p.accept(sourceFeature, f));
```

Comment:
Minor comment - when the lambda version isn't much shurter than the non-lambda I usually tend to prefer non-lambda for simplicity and to avoid some allocations

```java
for (var processor : attributeProcessors) {
  processor.accept(sourceFeature, processor);
}
```